enhancements/sig-architecture/32-extensiblescheduling/32-extensiblescheduling.md

```diff
@@ -0,0 +1,462 @@
+	// +optional
+	Conditions []metav1.Condition `json:"conditions"`
+
+	// Scores contains a list of score name and value of this managed cluster.
+	// +optional
+	Scores []AddOnPlacementScoreItem `json:"scores,omitempty"`
+
+	// ValidUntil defines the valid time of the scores.
+	// After this time, the scores are considered to be invalid by placement. nil means never expire.
+	// The controller owning this resource should keep the scores up-to-date.
+	// +optional
+	// +kubebuilder:validation:Required
+	// +kubebuilder:validation:Type=string
+	// +kubebuilder:validation:Format=date-time
+	ValidUntil *metav1.Time `json:"validUntil" protobuf:"bytes,4,opt,name=lastTransitionTime"`
+}
+
+//AddOnPlacementScoreItem represents the score name and value.
+type AddOnPlacementScoreItem struct {
+	// Name is the name of the score
+	// +kubebuilder:validation:Required
+	// +required
+	Name string `json:"name"`
+
+	// Value is the value of the score. The score range is from -100 to 100.
+	// +kubebuilder:validation:Minimum:=-100
+	// +kubebuilder:validation:Maximum:=100
+	// +required
+	Value int32 `json:"value,omitempty"`
+}
+
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// AddOnPlacementScoreList is a collection of AddOnPlacementScore.
+type AddOnPlacementScoreList struct {
+	metav1.TypeMeta `json:",inline"`
+	// Standard list metadata.
+	// More info: https://git.k8s.io/community/contributors/devel/api-conventions.md#types-kinds
+	// +optional
+	metav1.ListMeta `json:"metadata,omitempty"`
+
+	// Items is a list of AddOnPlacementScore
+	Items []AddOnPlacementScore `json:"items"`
+}
+```
+### Placement API
+Replace `Name` with `PrioritizerScoreCoordinate` inside `PrioritizerConfig`, to support specifying the customized score to sort clusters.
+
+```golang
+type Placement struct {
+  ...
+	Spec PlacementSpec `json:"spec"`
+}
+
+type PlacementSpec struct {
+  ...
+
+	// PrioritizerPolicy defines the policy of the prioritizers.
+	// If this field is unset, then default prioritizer mode and configurations are used.
+	// Referring to PrioritizerPolicy to see more description about Mode and Configurations.
+	// +optional
+	PrioritizerPolicy PrioritizerPolicy `json:"prioritizerPolicy"`
+}
+
+// PrioritizerPolicy represents the policy of prioritizer
+type PrioritizerPolicy struct {
+  ...
+
+	// +optional
+	Configurations []PrioritizerConfig `json:"configurations,omitempty"`
+}
+
+// PrioritizerConfig represents the configuration of prioritizer
+type PrioritizerConfig struct {
+	// PrioritizerScoreCoordinate represents the configuration of the prioritizer and score source.
+	// +kubebuilder:validation:Required
+	// +required
+	PrioritizerScoreCoordinate PrioritizerScoreCoordinate `json:"scoreCoordinate"`
+
+	// Weight defines the weight of the prioritizer score. The value must be ranged in [0,10].
+	// Each prioritizer will calculate an integer score of a cluster in the range of [-100, 100].
```

Comment:
prioritizing the clusters by simply ordering by their weight/score will be leading to "frequent-order-flipping" issues IMHO. e.g. if the scores are calculated from the metrics by an external system, mostly the metrics value will be actively changing in a range over time even if everything runs stably/normally --- say we have a policy that selects 2 clusters out of 3, and all these clusters' score are ranging in a "narrow" range e.g. [50,51] over time, in this case the prioritized order of the 3 clusters will be easily flipping which in the end makes the higher systems (e.g. a multi-cluster workload) working on that policy receiving too frequent topology update from the placement decision. 
 
in addition to that, i can also think of a cyclic-consequence problem if the clusters are merely prioritized  by their rank --- e.g. again if we select a cluster from the candidate clusters by the scores, then the score of the selected cluster can be descending b/c of being selected (e.g. if the score is calculated from the allocated resource), then the next moment this cluster can be kicked from the list b/c a candidate w/ higher score is now winning. repeatedly the list of the selected cluster will also be frequently changing over time.

so i suppose we can probably introduce a  `"margin"` concept into the prioritzer. a margin is a buffer for preventing the selected cluster list to be frequently flipping. hence, overall the clusters are ordered by their score but as for the candidate clusters, they will have to win the selected cluster $margin more scores to be chosen by the prioritizer. if the scores are ranged in [-100,100] so a reasonable/recommended value of margin can be 20 i suppose.

Reply:
@yue9944882 yes, we also have similar concerns, what we have done so far is to introduce a prioritizer called "[Steady](https://github.com/open-cluster-management-io/placement/blob/main/pkg/plugins/steady/steady.go)",  try to make the decision steady when other score changes. And user can [adjust the Steady weight](https://github.com/open-cluster-management-io/enhancements/tree/main/enhancements/sig-architecture/15-resourcebasedscheduling#4-pin-the-placementdecisions-use-story-4) in placement. It looks like similar to what you think.